src/commands/compose/getComposeServiceList.ts

```diff
@@ -0,0 +1,56 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { IActionContext, IAzureQuickPickItem } from 'vscode-azureextensionui';
+import { ext } from '../../extensionVariables';
+import { localize } from '../../localize';
+import { execAsync } from '../../utils/spawnAsync';
+
+const composeCommandReplaceRegex = /(up|down).*$/i;
+
+export async function getComposeServiceList(context: IActionContext, workspaceFolder: vscode.WorkspaceFolder, composeCommand: string): Promise<string> {
+    const services = await getServices(workspaceFolder, composeCommand);
+
+    // Fetch the previously chosen services list. By default, all will be selected.
+    const workspaceServiceListKey = `vscode-docker.composeServices.${workspaceFolder.name}`;
+    const previousChoices = ext.context.workspaceState.get<string[]>(workspaceServiceListKey, services);
+
+    const pickChoices: IAzureQuickPickItem<string>[] = services.map(s => {
+        return {
+            label: s,
+            data: s,
+            picked: previousChoices.some(p => p === s),
+        };
+    });
+
+    const subsetChoices =
+        await ext.ui.showQuickPick(
+            pickChoices,
+            {
+                canPickMany: true,
+                placeHolder: localize('vscode-docker.getComposeServiceList.choose', 'Choose services to start'),
+            }
+        );
+
+    context.telemetry.measurements.totalServices = pickChoices.length;
+    context.telemetry.measurements.chosenServices = subsetChoices.length;
+
+    // Update the cache
+    await ext.context.workspaceState.update(workspaceServiceListKey, subsetChoices.map(c => c.data));
+
+    return subsetChoices.map(c => c.data).join(' ');
+}
+
+async function getServices(workspaceFolder: vscode.WorkspaceFolder, composeCommand: string): Promise<string[]> {
+    // Start by getting a new command with the exact same files list (replaces the "up ..." or "down ..." with "config --services")
+    const configCommand = composeCommand.replace(composeCommandReplaceRegex, 'config --services');

```

Comment:
Will the regex work if `up` or `down` happen to show up elsewhere in the command line, such as in the compose filenames?

Reply:
No, great catch! I'll enforce word boundaries around those words.